### DISCUSSION TITLE
Composition by mole fractions

### DISCUSSION THREAD

**Discussion by dhoulek** [dgehringer/sqsgenerator#49] [Ideas]
Hi @dgehringer ,

I remember we removed the possibility of specifying composition by mole fractions as it can lead to non-deterministic results for some compositions/supercell sizes. I am wondering whether we can/should reimplement it back a possibly issue a user warning. Any strong opinion on this?

Do you foresee any principal difficulties when we would consider also several sublattices and will work with the mole fractions w.r.t. sublattice rather than overall composition? I mean the following:

(Ti0.25Al0.75) (B0.5N0.5) can be specified as 
* 0.25-0.75 on metal sublattice and 0.5-0.5 on the non-metal sublattice or
* 0.125-0.375-0.25-0.25 overall.
The latter for me makes sense only when we would consider full mixing, e.g. without sublattices, so I would be inclined to have the mole fractions always w.r.t. sublattice.

**Comment by dgehringer** [comment]
Hi @dhoulek 

Basically, from a technical point of view this if of course not a problem. There are two main reasons why I think we should keep it as it 

1. Specifying integer values you can uniquely identify each site. As you mentioned specifying fractional numbers that might lead to unwanted compositions. Furthermore rounding/parsing might lead to slightly different compositions. Simply consider 27 atoms *fcc* $3\times 3 \times 3$ with $A_{0.5}B_{0.5}$. Go for 13 A and 14 B or vice versa? $\rightarrow$ [see here](https://zen-of-python.info/in-the-face-of-ambiguity-refuse-the-temptation-to-guess.html#12)  
2. Adding a second way to specify composition might introduce confusion. There are already two ways to do it depending you want to populate a sub lattice only or the whole cell $\rightarrow$ [see here](https://zen-of-python.info/there-should-be-one-and-preferably-only-one-obvious-way-to-do-it.html#13) 

I am no per se against it? What would be the benefits besides saving a computing the number of atoms manually?

**Reply by dhoulek** [reply]
The motivation would be for different supercell sizes to keep the number of input changes to a minimum. I believe this could benefit the user experience.

Because of the ambiguity, I would propose that the code issues either a warning or stops completely in case the fractions would not lead to an integer number of sites.

**Reply by dgehringer** [reply]
Agree! 

If you confirm that it would be desirable, we can do so!

I'd just propose to fail with a hard stop in case of ambiguity. While a warning might be obvious for you, it might not be so obvious for other users, and I personally think unexpected behaviour (which it might be from another users perspective) is worse than failing.

**Reply by dgehringer** [reply]
I have created an issue https://github.com/dgehringer/sqsgenerator/issues/51 for this suggestion!

**Comment by dgehringer** [comment]
In contrast what do you think about **truly** allowing for a multiple sub lattice optimisation. Because the [current way of doing it](https://sqsgenerator.readthedocs.io/en/latest/advanced_topics.html#multiple-independent-sublattices-from-text-tin-rightarrow-left-text-ti-0-25-text-al-0-25-right-left-text-b-0-25-text-n-0-25-right) is not 100% clean. 

Especially setting the prefactor mode with `mul` is a pretty bad and tacky workaround which just works for this constructed example, and a few limited cases!

My idea goes something like this!

- I'd start with the requirement of mutual exclusive sublattices (SL)
  - each site belongs to exactly one SL
  - the SLs do not interact with each other 

 - Given $\Delta$ sublattices, one would not not only compute the SRO parameter $\alpha^i_{\xi\eta}$ for each species pair and each shell but also for each sub lattice
   - we would have to add one dimension to the SRO parameter $\alpha^{I,\Delta}_{\xi\eta}$
   - Note that the set of species might be different for each sub lattice, hence the size of the $\alpha^{I,\Delta}_{\xi\eta}$ tensor varies for each SL $\Delta$ which hence becomes a 4D
    
 - then compute the Objective function individually for each SL and optimise for a weighted (by the number of atoms on the sub lattice) sum of the objective function $O(\sigma) = \sum_\Delta \frac{N^\Delta}{N} O^{\Delta}(\sigma)$
 - then you the the refactors correctly which is hard to do with the current set up
 
Does this make sense? Do you think that this would be used?

**Reply by dhoulek** [reply]
This makes sense and will hugely simplify the user experience. We would only miss the possibility to control interactions between sublattices. However, I am not sure if this ever anyone used (I didn't) and also it's physical meaning may be debatable. For that, we should already include the real interatomic interactions.

I like the proposed weighting of the total objective function based on the number of sublattice sites. Maybe this can be the default behaviour, but the user can also specify other values if (s)he wants?

Just as a side comment, I don't think $\alpha_{\xi,\eta}^{I,\Delta}$ are tensors, so maybe to be on the safe side, let's call them just 4D matrices (or 4-index matrices).
